From c760a0ecf7125b66605d666350b3d4f84f35b7a3 Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Tue, 30 Jul 2024 18:33:31 +0800
Subject: [PATCH] WLANRTSYS-83617 Do Cleanup of Same STA MLD-MAC in all
 interfaces

Issue:
If a connected MLD STA tries to associate in other MLO VAP in the same AP
with the same Link Mac, the pruning of the same MLD MAC is not happening
since there is no sta entry for the STA in new bss. As a result, SID request
with existing mac address will be sent to Mac2Index module and fatal occurs.

Fix:
If the associating station is ML STA, do the Same STA MLD-MAC cleanup in
all interfaces in authentication phase.

FIXUP to WLANRTSYS-62845 AP MLD Auth Assoc and eapol changes
---
 src/ap/ieee802_11.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 19de0f8cc..25673d1cb 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3136,8 +3136,6 @@ static void handle_auth(struct hostapd_data *hapd,
 	sta = ap_get_sta(hapd, sa);
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	if (sta && (sta->flags & WLAN_STA_ASSOC)) {
-		mxl_auth_parse_ml_ie(hapd, mgmt, len, sta);
-		mxl_hostapd_cleanup_same_mld_mac_sta(hapd, sta);
 		mxl_hostapd_drv_cleanup_sta(hapd, sta);
 		sta = NULL;
 	}
@@ -3199,6 +3197,11 @@ static void handle_auth(struct hostapd_data *hapd,
 			resp = WLAN_STATUS_AP_UNABLE_TO_HANDLE_NEW_STA;
 			goto fail;
 		}
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+		mxl_auth_parse_ml_ie(hapd, mgmt, len, sta);
+		if (sta->mxl_sta_info.ext_flags & WLAN_STA_MULTILINK)
+			mxl_hostapd_cleanup_same_mld_mac_sta(hapd, sta);
+#endif
 	}
 
 #ifdef CONFIG_IEEE80211BE
-- 
2.43.0

